platform-sdk/docs/proposals/metric-labels/README.md

```diff
@@ -0,0 +1,341 @@
+Labels are key-value pairs that can be used to filter and group metrics.
+Why this proposal uses the name "metric labels" (like prometheus does) other libraries (like micrometer or open telemetry) use the term tags or attributes.
+
+## Motivation
+
+Today the metrics module only supports the definition of metrics without adding custom labels.
+While each metric has a hardcoded category, name, and nodeId no other custom information can be added.
+This proposal suggests adding a new concept called "metric labels" to the metrics module and
+migrate the `category`, `name`, and `nodeId` information to labels.
+While the category and name are defined as String values the nodeId is a class called `NodeId`.
+By doing so the metrics module becomes more flexible and can be used 100% independent of the
+platform module since it does not depend on the platform module's `nodeId` anymore.
+
+## Goals
+
+- Add a new concept called "metric labels" to the metrics module
+- Migrate the `nodeId` information to a label
+- Make the metrics module independent of the platform module
+- Provide a public API to define labels for metrics
+- Add labels to the snapshot API
+- The output of the snapshot API should contain the labels
+- The output of the snapshot API must be compatible with the current output (no breaking changes)
+- Create better and more dynamic dashboards in Grafana by using labels
+
+## Non-Goals
+
+Even with the new concept the metrics api should be only used in our (hashgraph) code.
+The metrics module should not be used by the end user.
+Here the preferred solution is still that we forward our metrics to a monitoring system that the custom application code uses.
+
+## Design & Architecture
+
+This chapter describes the changes that are needed to add labels to the metrics module.
+
+### About labels
+
+Before we start with the implementation we need to define what labels are.
+Labels are key-value pairs that are used to define, filter and group metrics.
+Labels are used by monitoring systems like Grafana to create more dynamic dashboards.
+By adding labels to the metrics module a single metric is now uniquely defined by its name and the labels.
+For example, metrics that counts the number of requests can have a label called `method` that defines if the request was a `GET` or `POST` request.
+That metrics are defined by the name "api_http_requests_total" and the label `method`.
+By doing so the monitoring system can create a dashboard that shows the number of requests per method.
+In the given example we would have 2 metric instances: one for `GET` and one for `POST`:
+
+- Metric 1: `api_http_requests_total{method="GET"}`
+- Metric 2: `api_http_requests_total{method="POST"}`
+
+While today the name of a metric is the unique identifier of a metric the name and the labels are the unique identifier of a metric when using labels.
+By doing so both metrics can use the same name but are still unique as long as the labels are different.
```

Comment:
Is this desirable? It seems quite confusing to me. Why not keep the unique identifier the same as before?

For example, it seems confusing to allow two metrics with the name `api_http_requests_total' with labels `method` and `handler` but allow another metric to be called `api_http_requests_total` with only the label `method`. To me, this should not be allowed.